### PR TITLE
[Bugfix] Fix Issue 828

### LINKED ISSUE #828: [Question] D-PACE loss denominator after the DFlash2 integration
Hi, I am the author of D-PACE. Thank you for integrating D-PACE into SpecForge and extending it to DFlash2.

While reviewing the current implementation, I noticed that the D-PACE loss is normalized by the sum of its position weights:

```python
loss = (neg_log_q * loss_weights).sum() / loss_weights.sum()
```

D-PACE position weights are derived from expected accepted length. Dividing the weighted loss by their sum cancels the total credit mass, which may change the intended relationship between the loss and expected accepted length.

In my earlier DFlash1 experiments, this normalization appeared to affect mean accepted length, although I no longer have the logs or checkpoints needed to report the exact numbers. I have also not yet tested whether the same effect occurs with DFlash2.

Would it be possible to normalize the weighted loss by the number of valid sampled anchors instead? Would this remain compatible with the shared DFlash/DFlash2 training path?

```python
loss = weighted_loss_sum / valid_anchor_count
```

### PR DESCRIPTION
**Pull request by jiapingW**
## Motivation

  The D-PACE objective was normalized by the sum of its position weights:

$$ L = \frac{\sum_{b,i} w_{b,i}\ell_{b,i}}{\sum_{b,i} w_{b,i}} $$

  Because D-PACE position weights represent expected accepted-length credit, dividing by their sum removes the overall credit mass and changes the intended objective.

  This PR changes the denominator to use sequence-balanced valid-anchor normalization. It preserves D-PACE credit while preventing sequences with more sampled anchors
  from receiving disproportionately larger weight:

  $$L =\frac{1}{B_{\mathrm{valid}}}\sum_b  \frac{\sum_i w_{b,i}\ell_{b,i}}{A_b}$$
  where $\(A_b\)$ is the number of valid sampled anchors in sequence $\(b\)$, and $\(B_{\mathrm{valid}}\)$ is the number of sequences containing at least one valid anchor.

  ## Modifications

  - Normalize D-PACE losses by valid sampled anchors instead of the sum of D-PACE position weights.
  - Balance each sequence by scaling its objective terms by $$\(1 / A_b\)$$.
  - Preserve additive numerator/denominator semantics across:
    - objective chunking
    - gradient accumulation
    - distributed data-parallel reduction
  - Apply the same per-sequence scaling to:
    - D-PACE CE
    - D-PACE with TV/LK objectives
    - cumulative-confidence-only and continuation-value-only ablations
    - the DFlash2 selector objective
  - Align detailed DFlash2 loss-weight metrics with the effective training weights.
  - Keep the standard `loss_type=dflash` behavior unchanged.
  - Add regression tests for:
    - batches with unequal valid-anchor counts
    - D-PACE CE and TV/LK variants
    - D-PACE ablations
    - zero-credit edge cases
    - DFlash2 selector scaling
    - chunked and unchunked objective equivalence
    - trainer accumulation and reduction behavior

  ## Related Issues

  Fixes #828

  ## Accuracy Test

  This PR changes the D-PACE loss normalization but does not modify the model architecture, inference path, or kernels.

  Formula-level and model-side regression tests pass:

  - 106 tests passed
  - 50 subtests passed
  - DFlash and DFlash2 objective tests passed
  - Chunked and unchunked loss/gradient equivalence tests passed
  - Trainer accumulation and reduction tests passed

  A full checkpoint training comparison is not included in this PR.

  ## Benchmark & Profiling
## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.io/docs/developer_guide/contribution_guide).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.io/docs/developer_guide/contribution_guide).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.io/docs/developer_guide/contribution_guide).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.io/docs/developer_guide/benchmark_and_profiling) and [Accuracy Results](https://docs.sglang.io/docs/developer_guide/evaluating_new_models).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by jiapingW** [comment]
@Lucas-TY, you can review it.

**Review by Lucas-TY — APPROVED** [review]
Yeah, I think it looks good to me. Thanks for fixing this!

**Comment by jiapingW** [comment]
@Lucas-TY If you think this is OK, you can approve the PR.

File: specforge/algorithms/common/dflash_family_model.py [hunk 1]
```diff
@@ -740,7 +754,17 @@ def _dflash_objective_chunk_terms(
                     self.loss_type,
                 )
             loss_weights = weight_mask * dpace_weights
-            loss_den = loss_weights.sum()
+            valid_anchors = (weight_mask > 0).any(dim=-1)
+            if sequence_anchor_scale is None:
```
Comment (maocheng23) [inline specforge/algorithms/common/dflash_family_model.py]: Could we require an explicitly provided sequence_anchor_scale in the D-PACE branch and raise a ValueError when it is missing?
The current forward() correctly computes the scale over the full sequence before chunking. However, this fallback can silently normalize by the chunk-local anchor count if a caller passes a partial chunk without the scale. Requiring it explicitly would make the full-sequence normalization contract clear.
We should apply the same constraint to the metrics path and update direct-call tests to pass the precomputed scale. A regression test for the missing-scale error would also help protect this contract.

Comment (jiapingW) [reply specforge/algorithms/common/dflash_family_model.py]: I have updated it.

**Review by Lucas-TY — APPROVED** [review]
I think it looks good

**Review by Lucas-TY — APPROVED** [review]
LGTM
